### ISSUE TITLE
Randomize tiles

### ISSUE THREAD

**Issue by pablogila** [pablogila/TileMapDual#13]
Add an option to place tiles from different atlases randomly, to add some variation.
This one should be easy to implement.

**Comment by maudrid** [comment]
Would this be the same as having 2 tiles that have the same terrain configuration, setting a high probability on one and a low one on the other and then painting the terrain would sometimes choose a different tile?
The "From different atlasses" part is confusing me.

The behaviour I described above is working on the TileMapLayer but not on the TileMapDual.

**Comment by pablogila** [comment]
Oh okay I wasn't aware of this feature. Is this using the same atlas?

**Comment by maudrid** [comment]
Here is the terrain configuration (these tiles are all individual tiles with completely filled terrain masks) :
![Image](https://github.com/user-attachments/assets/45106567-76a0-441c-a3df-f9dc84f9ee87)

Here is how I've set up the probabilities:
![Image](https://github.com/user-attachments/assets/8e45c487-2960-4f62-918e-76a198485105)

As you can see in both screenshots, it's only using the last tile on the far right.
If I set the same thing up in a TileMapLayer, it works as expected. 
Like this:

![Image](https://github.com/user-attachments/assets/64dfc61d-1764-4551-b3a4-024e012e589e)

**Comment by pablogila** [comment]
Thanks for the clarification! I see a problem here. TileMapDual computes all the cells every time it is loaded, so the tiles would be randomized every time the player enters in the scene. This is not ideal, and I don't see any way around it, aside from just manually painting the different tiles on top of the map... Any suggestion here is welcome!

**Comment by maudrid** [comment]
If I understood how TileMapDual works under the hood I bit better, I could try to help out.
Do you have any technical documentation about the architecture I can read?

**Comment by pablogila** [comment]
That would be pretty much appreciated :D

This is a scheme of TileMapDual under the hood:

![](https://raw.githubusercontent.com/pablogila/TileMapDual/refs/heads/main/docs/v5_structure.png)

There are also descriptions embedded into the code.

I think it makes sense to automate the compilation of a full documentation page, kinda like Pdoc or Sphynx for Python, cause otherwise a manual documentation gets quickly outdated. I will look into it...

**Comment by maudrid** [comment]
Hi.
I wanted to check if I'm on the right track.
It seems to me that this code is responsible for deciding what tile to display:
DisplayLayer.gd ->
```
## Updates a specific world cell.
func update_tile(cache: TileCache, cell: Vector2i) -> void:
	var get_cell_at_path := func(path): return cache.get_terrain_at(follow_path(cell, path))
	var terrain_neighbors := _terrain.display_to_world_neighborhood.map(get_cell_at_path)
	var mapping: Dictionary = _terrain.apply_rule(terrain_neighbors)
	var sid: int = mapping.sid
	var tile: Vector2i = mapping.tile
	set_cell(cell, sid, tile)
```

Specifically where it calls: `var mapping: Dictionary = _terrain.apply_rule(terrain_neighbors)`
The mapping contains the tile that should be applied to the cell.
What confuses me a bit is the comment on the apply_rule function:
`## Returns the tile that should be used based on the surrounding terrain neighbors`
This seems to indicate that the apply_rule function does not do anything to the current cell, but on the cells that surround that cell.

If I'm correct that this is the code responsible for selecting the tile to apply to the cell, can I ask why you wrote the custom implementation instead of using the built in functionality of the TileMapLayer? It's seems to do a good job with selecting the correct tile according to the terrain setup.

**Comment by pablogila** [comment]
TerrainLayer.apply_rule()` takes the neighbors of a given cell and decides which tile to display for that cell based on those neighbors. It does not *do* anything to the neighbors, it just checks them to decide which tile to display according to certain rules that depend on the grid type.

I ignore if there is a native way to apply these grid dependent rules in TileMapLayer... Of course, if you know how to do it, please let us know!

**Comment by maudrid** [comment]
Great, so it seems like TerrainLayer.apply_rule() is where I should be focusing my effort.
Thanks.
PS: I only work on this during weekends, so it may take a while, sorry.

**Comment by maudrid** [comment]
I spent today looking at this issue.
The TileMapLayer code for choosing the tiles will not work for this.
My idea was to use that, but in the TileMapDual layer, it effectively uses 4 (corner) tiles when you place a single tile down, while this is not the case with the TileMapLayer.
So we cannot use the same algorithm.
I reviewed what I need for my own game and I think my game is simple enough to use the normal TileMapLayer with only slightly adjusted tiles and a custom data layer to identify what tiles I can use instead of using the terrain itself.
So sadly I won't be continuing on this issue, sorry for that.
Maybe I'll come back to this if it turns out that the TileMapDual is better for my game after all.

**Comment by Illauriel** [comment]
> This is not ideal, and I don't see any way around it, aside from just manually painting the different tiles on top of the map... Any suggestion here is welcome!

The randomization needs to be deterministic so that every "reroll" produces the same results. I'll try looking into it ~

UPD: Guess I solved it

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/2ed8145b-31b1-4e6d-91cb-2e96aad04195" />
<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/e3a887cf-fdd5-441a-93e7-3b86d302d2b3" />

Basically in `_register_tile()` I save probablilty into mapping, and in `_register_rule()` instead of storing one rule, and firing a warning if there's already one, I changed the method to save it as an array of mappings. At `apply_rule()` time we do a weighted roll, based on saved probabilities, and voila.

Should I submit it as a PR?

**Comment by pablogila** [comment]
This is a very interesting approach, thanks for sharing! Feel free to submit a PR  :D
